text/0000-bench-utils.md

```diff
@@ -0,0 +1,157 @@
+- Feature Name: black_box-and-clobber
+- Start Date: 2018-03-12
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC adds two functions to `core::mem`: `black_box` and `clobber`, which are
+mainly useful for writing benchmarks.
+
+# Motivation
+[motivation]: #motivation
+
+The `black_box` and `clobber` functions are useful for writing synthetic
+benchmarks where, due to the constrained nature of the benchmark, the compiler
+is able to perform optimizations that wouldn't otherwise trigger in practice.
+
+The implementation of these functions is backend-specific and requires inline
+assembly. Such that if the standard library does not provide them, the users are
+required to use brittle workarounds on nightly.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+## `mem::black_box`
+
+The function:
+
+```rust
+pub fn black_box<T>(x: T) -> T;
+```
+
+prevents the value `x` from being optimized away and flushes pending reads/writes
+to memory. It does not prevent optimizations on the expression generating the
+value `x` nor on the return value of the function. For
+example ([`rust.godbolt.org`](https://godbolt.org/g/YP2GCJ)):
+
+```rust
+fn foo(x: i32) -> i32{ 
+  mem::black_box(2 + x);
+  3
+}
+let a = foo(2);
+```
+
+Here, the compiler can simplify the expression `2 + x` into `2 + 2` and then
+`4`, but it is not allowed to discard `4`. Instead, it must store `4` into a
+register even though it is not used by anything afterwards.
+
+## `mem::clobber`
+
+The function
+
+```rust
+pub fn clobber() -> ();
+```
+ 	 
+flushes all pending writes to memory. Memory managed by block scope objects must
+be "escaped" with `black_box` . 
+
+Using `mem::{black_box, clobber}` we can benchmark `Vec::push` as follows:
+
+```rust
+fn bench_vec_push_back(bench: Bencher) -> BenchResult {
+    let n = /* large enough number */;
+    let mut v = Vec::with_capacity(n);
+    bench.iter(|| {
+        // Escape the vector pointer:
+        mem::black_box(v.as_ptr());
```

Comment:
does the black box need to be inside each iteration?

Reply:
Not in this case because the vector does not grow. If the vector would be allowed to grow on push, then the answer would be yes.